bin/utils.js

```diff
@@ -142,17 +142,22 @@ class WSSharedDoc extends Y.Doc {
  *
  * @param {string} docname - the name of the Y.Doc to find or create
  * @param {boolean} gc - whether to allow gc on the doc (applies only when created)
- * @return {WSSharedDoc}
+ * @return {Promise<WSSharedDoc>}
  */
-const getYDoc = (docname, gc = true) => map.setIfUndefined(docs, docname, () => {
+const getYDoc = async (docname, gc = true) => {
+  let set = map.get(docname)
+  if (set !== undefined) {
+    return set
+  }
+
   const doc = new WSSharedDoc(docname)
   doc.gc = gc
   if (persistence !== null) {
-    persistence.bindState(docname, doc)
+    await persistence.bindState(docname, doc)
   }
   docs.set(docname, doc)
```

Comment:
This doesn't look right. We set `doc` in `docs`, but `map.setIfUndefined` also sets doc in `docs`.

You did right in removing the `map.setIfUndefined` part, but by making it asynchronous you don't handle the case that the document is created concurrently. I.e. if you call getYDoc twice you always create two different documents.

I think the right approach is to ..

1. Call map.setIfUndefined as before. getYDoc should be a synchronous function.
2. Define a property `doc.whenSynced = persistence.bindState(docname, doc)` on WSSharedDoc. 
3. In `setupWSConnection` you call `await doc.whenSynced` before syncing with the client.

Reply:
Thanks for the quick feedback, I certainly don't want to create _another_ race condition. Will work on this approach